### DISCUSSION TITLE
Custom JWT Claims / Dynamic Variables

### DISCUSSION THREAD

**Discussion by ernst86913** [directus/directus#17138] [Feature Requests]
Hi Everyone,
having the ability to customize token claims is really great.

What I am missing is how to use those claims as part of the permission system… that’s when those custom claims really make sense.

example:  $JWT.xxxxxx 

This would make the current permission system much more flexible / powerful. 

Is this something you are considering?

**Comment by rijkvanzanten** [comment]
From a permissions standpoint, how is it different from adding custom fields to the user user or role and relying on $CURRENT_USER / $CURRENT_ROLE? 🤔

**Reply by ernst86913** [reply]
Hmmm - so your approach would be to add custom fields to user record - and every time I issue custom claims to a user I update the custom user field(s) to then query it in the permissions….

Should work - though not super nice - but yes - should be a good workaround.

Wouldn’t it be better in general to take ANY dynamic permission variables from the JWT token? JWT.user, JWT.role, JWT.custom1, JWT.custom2,… ?

anyway - thanks for your thoughts - I think I can make it happen using your approach

**Reply by rijkvanzanten** [reply]
How would you add the custom claims to the token? Through a hook or..? Just trying to wrap my head around how this flow would ideally look like from your perspective 🙂

**Reply by ernst86913** [reply]
The hook for the custom claims already exists in Directus… “auth.jwt”

It’s even possible to customize the accountability object using the hook “authenticate”

Only thing which is missing in permissions is to use those custom variables (either from auth.jwt or from accountability object) as dynamic variables in permission expressions… both options would work nicely

**Reply by ernst86913** [reply]
Maybe easier to have a use case:  I’d like to have organizational units (legal entities) which need to be strictly separated from each other. Users usually are part of one unit… but some managers are responsible for multiple units.

For strict separation I’d like to use JWT which contains the org unit a user is in… and I have a custom endpoint which allows users to switch units if they are in multiple units. In this case the token is renewed with the org unit the user switched to.

Challenge 1:  permissions. All requests need to be verified for strict separation. As I already have the org unit in the JWT / accountability I thought it makes sense to use that info directly in the permission rules.

key question regarding your proposed $user.org approach: will this have any negative impacts on performance? Will a $user.org result in a DB query with each request? If that’s not the case I’d simply go with your recommended approach.

Challenge 2: ALL records need to have an org ID they belong to… initially I thought to make sure I add this with every create / delete / update request in nodejs by using one single hook with app.use and taking the org from the token… issue: this does not work if records are eg created „inside“ express server with direct service calls. Example: activity log. Means - I can use app.use hook - but for activities, revisions,… I need to hook into each of the operations individually…

Key question: any other smart idea how to solve this in a nicer way? Maybe with field presets? Or similar?

again - thanks a lot for your help. Really appreciated

**Reply by ernst86913** [reply]
Maybe best way to solve challenge2 is enforcing it via knex-hooks… will try wether this works

**Reply by fabiobeoni** [reply]
> From a permissions standpoint, how is it different from adding custom fields to the user user or role and relying on $CURRENT_USER / $CURRENT_ROLE? thinking

This kind of solution does work well. BTW think about the case of using an external IDP (eg.: Keycloak/OKTA) where you have already profiled the user-base with groups and roles. On the IDP side you can map such groups/roles to the JWT token, and then the proposal of accessing them from `$CURRENT_USER.JWT.xxxxx` sounds very solid (no repetitions, no risk of misalignment, etc...).

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
